### ISSUE TITLE
Allow passing of container log timestamp to fluentd log driver

### ISSUE THREAD

**Issue by bradbeam** [moby/moby#17181]
Currently there is a limitation in fluentd where it does not support millisecond timing ( https://github.com/fluent/fluentd/issues/461 ). This causes some nonsense when using an ELK like backend when searching/reviewing logs. Docker emits the timestamp to the fluent-logger-golang, but this timestamp is converted to unix time ( seconds ) to work within the bounds of fluentd. Since docker already has the precise timestamp, it'd be great if this could be passed along as part of the log message data that gets emitted to fluentd so some post processing could be done after fluentd. 

I'm guessing it'd involve adding a new key in the map, https://github.com/docker/docker/blob/master/daemon/logger/fluentd/fluentd.go#L77 to include the msg.Timestamp. 

Thoughts?

**Comment by ashb** [comment]
Fluentd v0.14 now has support for nanosecond timestamps, so we'd "just" need to change the msgpack we send to fluentd (possibly make it flagged).

**Comment by farmdawgnation** [comment]
@ashb It looks like Fluentd 0.14 has been released. Any chance this might land soon?

**Comment by ashb** [comment]
@farmdawgnation not from me as I'm just a lowly docker user I'm afraid.

**Comment by farmdawgnation** [comment]
Ah, sorry. I thought you were a maintainer. :)

**Comment by timka** [comment]
@thaJeztah What's the status? Thanks!

**Comment by thaJeztah** [comment]
@timka think this needs someone to open a pull request, perhaps @bradbeam is interested, or @tagomoris

**Comment by tagomoris** [comment]
My concern is, sending sub-second time breaks Fluentd v0.12. So I think it's better to release sub-second support of fluentd logging driver after stable release of Fluentd v0.14 (2016 Q4 or 2017 Q1).

**Comment by bradbeam** [comment]
@tagomoris would it be worth adding a log opt for fluentd to specify message format version ( v0.12 vs v0.14+ )? 

What are your thoughts about including that flag as part of the data map that is emitted to `fluentd.PostWithTime()` and then doing the time conversion to time.Unix() or time.UnixNano() (both return int64) based on that map value?

**Comment by evmin** [comment]
Any news on the case? Would be very much interested...

**Comment by shane-axiom** [comment]
Several pull requests on https://github.com/uken/fluent-plugin-elasticsearch were recently merged allowing it to handle sub-second timestamps, so it looks like Docker is the last missing piece in a Docker EFK log stack with sub-second precision. Would be great to see this get fixed.

**Comment by shane-axiom** [comment]
Also, if the intent is to add a log opt which passes sub-second times to fluentd, I suggest that this issue be renamed to something like:

"Add fluentd-nano-time log-opt to fluentd logging driver"

**Comment by shane-axiom** [comment]
In the meantime, if anyone knows of a workaround allowing a custom sub-second timestamp key to be passed to fluentd I'd love to hear it.

**Comment by cosmo0920** [comment]
Hi, I'm one of the fluent-plugin-elasticsearch plugin maintainer.
This plugin's v0.14 migration status is still work in progress.
Currently, I do not release final 2.0.0 version which support nanosecond precision time.
This is because v0.12 and v0.14 forward protocols do not have compatibility for nanosecond precision time handling.
In v0.12 just using unix epoch which is represented with Integer.
But in v0.14, Fluentd can handle unix epoch and EventTime, which is newly introduced ext type for nanosecond precision time.

ref: https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1#eventtime-ext-format 

In the mean time, sending sub second precision time should be configurable and disabled by default for compatibility against v0.12.
Is the above implementation possible?

**Comment by yissachar** [comment]
Fluentd v0.14 is now stable as of v0.14.22: https://www.fluentd.org/blog/fluentd-v0.14.22-has-been-released

**Comment by hai-ld** [comment]
So if my understanding is correct, currently Docker does not send timestamp with sub-second precision to Fluentd agent. Is there anyway to mitigate this problem?

Currently our system is shipping log through GELF to a central Graylog server, but we want to improve it to be more robust against network disruption, so we intend to have Fluentd agent on each host to act as buffered forwarder. This seems to satisfy our requirement on network disruption-proof, but without millisecond or finer precision, the central log server cannot sort the messages correctly.

**Comment by shane-axiom** [comment]
@hai-ld I haven't had time to try this myself, but [this comment](https://github.com/uken/fluent-plugin-elasticsearch/pull/223#issuecomment-335268826) from @wardviaene looks promising:

>The issue moby/moby#17181 still needs to be resolved before you'll get your messages from docker in sub-second format. The only solution now is to put a new timestamp on the incoming record and write it in subsecond format to ES
>
>timestamp_ms ${ require 'time'; Time.now.utc.iso8601(3) }
>time_key timestamp_ms
>
>You possibly also need time_precision and time_key_format

**Comment by shane-axiom** [comment]
@hai-ld Ok, my curiosity was piqued so I tried it out, and it appears to be working. My simplified fluent.conf below:

```
<source>
  @type forward
  port 24224
  bind 0.0.0.0
  source_hostname_key hostname
</source>

<filter **>
  @type record_transformer
  enable_ruby true
  <record>
    timestamp_ms ${ require 'time'; Time.now.utc.iso8601(3) }
  </record>
</filter>

<match **>
  @type copy
  <store>
    @type elasticsearch
    host elasticsearch
    port 9200
    logstash_format true
    time_key timestamp_ms
  </store>
</match>
```

**Comment by hai-ld** [comment]
Many thanks, @shane-axiom. Actually we're trying a solution similar to this on dev envronment, but we don't like it because:
- data duplication: now there're two timestamp fields)
- compatibility problem: current log messages have second-precision `timestamp` field, new log messages have second-precision `timestamp` and millisecond-precision `timestamp_ms` fields, and future log messages (when Docker supports Fluentd 0.14) will only have nano/millisecond-precision `timestamp` field). Currently we're using emsearcy/fluent-plugin-gelf so there's no `time_key` to avoid this problem.
- the timestamp is generated when fluentd agent receives the message, not when the message is created, so there might be some edge cases on production

**Comment by shane-axiom** [comment]
@hai-ld Yeah, it's definitely a kludge. The drawbacks you mentioned are acceptable to us as a workaround to have log messages in the correct order until this is fixed and released.

**Comment by hai-ld** [comment]
To anyone having this same problem with Docker-FluentD-Graylog stack, we solved it as follow:
- in filter, add a `msec` record
```
<filter docker.**>
  @type record_transformer
  enable_ruby
  <record>
    msec ${ require 'time'; Time.now.utc.strftime("%L") }
  </record>
</filter>
```
- in match, enable `add_msec_time`
```
    @type gelf
    host graylog_host
    port graylog_port
    ...
    add_msec_time true
```
The plugin https://github.com/emsearcy/fluent-plugin-gelf has no documentation whatsoever, so tracing down to this setup is so much harder than it needs to be, especially when we're not a Ruby shop :disappointed:

**Comment by scottbuettner** [comment]
Hey @tagomoris any chance we can get this moving forward? Now that fluentd 1.0 is out, I'd imagine getting millisecond or nanosecond times should be fine. Being limited to second precision is a real obstacle to our deployments currently. Thanks!

**Comment by tagomoris** [comment]
@scottbuettner Yes, it looks ok for me to move it forward.
Now I can't spend my time for it, but I'll agree to add a feature to enable `SubSecondPrecision` of fluent-logger-golang, and will review such change.

**Comment by shane-axiom** [comment]
It looks like this was quietly implemented in November 2017 [here](https://github.com/moby/moby/commit/a1ebda09bab5bf22d96a15e71618c4a0ac08bb55) by adding the [`fluentd-sub-second-precision` log-opt](https://docs.docker.com/config/containers/logging/fluentd/).

```
docker run --log-driver=fluentd --log-opt fluentd-address=fluentdhost:24224 \
   --log-opt fluentd-sub-second-precision=true ...
```

Looks like it has been available since docker-ce [v17.12](https://github.com/docker/docker-ce/blob/v17.12.0-ce/components/engine/daemon/logger/fluentd/fluentd.go) and [v18.01](https://github.com/docker/docker-ce/blob/v18.01.0-ce/components/engine/daemon/logger/fluentd/fluentd.go).

@bradbeam @thaJeztah I think this can be closed?

**Comment by cpuguy83** [comment]
Thanks, I believe you are right.
